Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp

```diff
@@ -104,6 +109,34 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
             g_cclosure_marshal_VOID__BOXED,
             G_TYPE_NONE, 1,
             WEBKIT_TYPE_JAVASCRIPT_RESULT);
+
+    /**
+     * WebKitUserContentManager::script-message-with-reply-received:
+     * @manager: the #WebKitUserContentManager
+     * @script_message_reply: the #WebKitScriptMessageReply holding the value received from the JavaScript world.
+     *
+     * This signal is emitted when JavaScript in a web view calls
+     * <code>window.webkit.messageHandlers.<name>.postMessage()</code>, after registering
+     * <code><name></code> using
+     * webkit_user_content_manager_register_script_message_handler_with_reply()
+     *
+     * The callback function is expected to send a reply that could be handled
+     * asynchronously by taking a reference of the reply object.
+     * If no reply is supplied by the user, an automatic reply with an
+     * undefined value will be sent.
+     *
+     * Since: 2.40
+     */
+    signals[SCRIPT_MESSAGE_WITH_REPLY_RECEIVED] =
+        g_signal_new(
+            "script-message-with-reply-received",
+            G_TYPE_FROM_CLASS(gObjectClass),
+            static_cast<GSignalFlags>(G_SIGNAL_RUN_LAST | G_SIGNAL_DETAILED),
+            0, g_signal_accumulator_true_handled, nullptr,
+            g_cclosure_marshal_generic,
+            G_TYPE_BOOLEAN, 1,
```

Comment:
G_TYPE_BOOLEAN -> G_TYPE_NONE
1 -> 2

Reply:
We need `G_TYPE_BOOLEAN` to be able to use `g_signal_accumulator_true_handled`, which allows a signal callback to claim that it has handled the signal and that the rest of connected callbacks should not run.

Given that a script message request can only be replied or errored *only once* (and the API functions assert that on entry), we don't want multiple callbacks trying to send replies, therefore we need this in order to stop signal emission and avoid e.g. multiple calls to `webkit_user_script_message_reply_return_value()`.

Reply:
(Changing `1`→`2` is still needed, of course, now that the signal has one more argument.)

Reply:
> true handled is useful when we have a default signal handler in the class, to prevent the default one from being invoked if already handled by the user. But user connecting twice to the same signal is just a programmer error, I would say.

I would use `g_signal_accumulator_true_handled` anyway. It's an extra level of safety and there's no downside.

Reply:
FWIW, I tend to favor APIs which follow the “provide mechanisms, not policies” paradigm, meaning that it tends to be better to allow people to decide how things better fit what their programs need, than to impose a strict policy on how things work&mdash;of course, we need to specify _some_ policies (like “always use the public embedding API from the same thread inside a `GMainLoop`”) because there may not be a way around them... but I think in this case we don't need to force a single way of doing things.

Reply:
E.g. Epiphany does this for the `decide-policy` signal; it handles different policy decision types in different places.

Reply:
You are both right, and now I've realized that we need to make it true handled to allow to handle the reply asynchronously. We should add the return value to the documentation and include that to handle it async you have to take a ref to the reply and return TRUE.